File: ProgressView/CircleProgressView.m
//
//  CircleProgressView.m
//  CircleProgressView
//
//  Created by Eric Rolf on 4/3/15.
//  Copyright (c) 2015 Eric Rolf. All rights reserved.
//

#import "CircleProgressView.h"

@interface CircleProgressView ()

struct Constants {              // Default Values = Private
    double circleDegress;       // = 360.0
    double minimumValue;        // = 0.000001
    double maximumValue;        // = 0.999999
    double ninetyDegrees;       // = 90.0
    double twoSeventyDegrees;   // = 270.0
    
};

@property (nonatomic, readwrite)           double internalProgress;
@property (nonatomic, readwrite)    struct Constants constants;

@end

@implementation CircleProgressView

@synthesize progress = _progress;

- (instancetype)init {
    if (self = [super init]) {
        [self setup];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super initWithCoder:aDecoder]) {
        [self setup];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        [self setup];
    }
    return self;
}

#pragma mark - Interface Builder

- (void)prepareForInterfaceBuilder {
    self.trackBackgroundColor = (self.trackBackgroundColor)? :[UIColor darkGrayColor];
    self.trackFillColor = (self.trackFillColor)? :[UIColor purpleColor];
    self.trackWidth = (self.trackWidth)? :10;
    self.clockwise = (self.clockwise)? :true;
    self.progress = (self.progress)? :.45;
    self.centerFillColor = (self.centerFillColor)? : [UIColor whiteColor];
}

#pragma mark - Setup

- (void)setup {
    self.contentView = [UIView new];
    self.constants = [self initConstants];
    
    [self addSubview:self.contentView];
}

- (struct Constants)initConstants {
    struct Constants constants;
    constants.circleDegress = 360.0;
    constants.minimumValue = 0.000001;
    constants.maximumValue = 0.999999;
    constants.ninetyDegrees = 90.0;
    constants.twoSeventyDegrees = 270.0;
    return constants;
}

#pragma mark - Setters

- (void)setProgress:(double)progress {
    _internalProgress = progress;
    _progress = progress;
    [self setNeedsDisplay];
}

- (void)setClockwise:(BOOL)clockwise {
    _clockwise = clockwise;
    [self setNeedsDisplay];
}

- (void)setTrackWidth:(CGFloat)trackWidth {
    _trackWidth = trackWidth;
    [self setNeedsDisplay];
}

- (void)setTrackImage:(UIImage *)trackImage {
    _trackImage = trackImage;
    [self setNeedsDisplay];
}

- (void)setTrackBackgroundColor:(UIColor *)trackBackgroundColor {
    _trackBackgroundColor = trackBackgroundColor;
    [self setNeedsDisplay];
}

- (void) setTrackFillColor:(UIColor *)trackFillColor {
    _trackFillColor = trackFillColor;
    [self setNeedsDisplay];
}

- (void)setTrackBorderColor:(UIColor *)trackBorderColor {
    _trackBorderColor = trackBorderColor;
    [self setNeedsDisplay];
}

- (void)setTrackBorderWidth:(CGFloat)trackBorderWidth {
    _trackBorderWidth = trackBorderWidth;
    [self setNeedsDisplay];
}

- (void)setCenterFillColor:(UIColor *)centerFillColor {
    _centerFillColor = centerFillColor;
    [self setNeedsDisplay];
}

- (void)drawRect:(CGRect)rect {
    [super drawRect:rect];
    
    CGRect innerRect = CGRectInset(rect, self.trackBorderWidth, self.trackBorderWidth);
    
    self.internalProgress = (self.internalProgress/1.0) == 0.0 ? self.constants.minimumValue : self.progress;
    self.internalProgress = (self.internalProgress/1.0) == 1.0 ? self.constants.maximumValue : self.internalProgress;
    self.internalProgress = self.clockwise ?
                            (-self.constants.twoSeventyDegrees + ((1.0 - self.internalProgress) * self.constants.circleDegress)) :
                            (self.constants.ninetyDegrees - ((1.0 - self.internalProgress) * self.constants.circleDegress));
    
    CGContextRef ctx = UIGraphicsGetCurrentContext();
    
    // background Drawing
    [self.trackBackgroundColor setFill];
    UIBezierPath *circlePath = [UIBezierPath bezierPathWithOvalInRect: CGRectMake(CGRectGetMinX(innerRect), CGRectGetMinY(innerRect), CGRectGetWidth(innerRect), CGRectGetHeight(innerRect))];
    [circlePath fill];
    
    if (self.trackBorderWidth > 0) {
        circlePath.lineWidth = self.trackBorderWidth;
        [self.trackBorderColor setStroke];
        [circlePath stroke];
    }
    
    // progress Drawing
    UIBezierPath *progressPath = [UIBezierPath new];
    CGRect progressRect = CGRectMake(CGRectGetMinX(innerRect), CGRectGetMinY(innerRect), CGRectGetWidth(innerRect), CGRectGetHeight(innerRect));
    CGPoint center = CGPointMake(CGRectGetMidX(progressRect), CGRectGetMidY(progressRect));
    CGFloat radius = progressRect.size.width / 2.0;
    CGFloat startAngle = self.clockwise ? (-self.internalProgress * M_PI / 180.0) : (self.constants.twoSeventyDegrees * M_PI / 180);
    CGFloat endAngle = self.clockwise ? (self.constants.twoSeventyDegrees * M_PI / 180) : (-self.internalProgress * M_PI / 180.0);
    
    [progressPath addArcWithCenter:center radius:radius startAngle:startAngle endAngle:endAngle clockwise:!self.clockwise];
    [progressPath addLineToPoint:CGPointMake(CGRectGetMidX(progressRect), CGRectGetMidY(progressRect))];
    [progressPath closePath];
    
    CGContextSaveGState(ctx);
    
    [progressPath addClip];
    
    if (self.trackImage) {
        [self.trackImage drawInRect:innerRect];
    } else {
        [self.trackFillColor setFill];
        [circlePath fill];
    }
    
    CGContextRestoreGState(ctx);
    
    // center Drawing
    UIBezierPath *centerPath = [UIBezierPath bezierPathWithOvalInRect:CGRectMake(CGRectGetMinX(innerRect) + self.trackWidth, CGRectGetMinY(innerRect) + self.trackWidth, CGRectGetWidth(innerRect) - (2 * self.trackWidth), CGRectGetHeight(innerRect) - (2 * self.trackWidth))];
    [self.centerFillColor setFill];
    [centerPath fill];
    
    CAShapeLayer *layer = [CAShapeLayer layer];
    layer.path = centerPath.CGPath;
    self.contentView.layer.mask = layer;
}

@end

File: ProgressView/CircleProgressView.swift
//
//  CircleProgressView.swift
//
//
//  Created by Eric Rolf on 8/11/14.
//  Copyright (c) 2014 Eric Rolf, Cardinal Solutions Group. All rights reserved.
//

import UIKit

@objc @IBDesignable open class CircleProgressView: UIView {

    fileprivate struct Constants {
        let circleDegress = 360.0
        let minimumValue = 0.000001
        let maximumValue = 0.999999
        let ninetyDegrees = 90.0
        let twoSeventyDegrees = 270.0
        var contentView:UIView = UIView()
    }

    fileprivate let constants = Constants()
    fileprivate var internalProgress:Double = 0.0

    fileprivate var displayLink: CADisplayLink?
    fileprivate var destinationProgress: Double = 0.0

    @IBInspectable open var progress: Double = 0.000001 {
        didSet {
            internalProgress = progress
            setNeedsDisplay()
        }
    }

    @IBInspectable open var refreshRate: Double = 0.0 {
        didSet { setNeedsDisplay() }
    }

    @IBInspectable open var clockwise: Bool = true {
        didSet { setNeedsDisplay() }
    }

    @IBInspectable open var trackWidth: CGFloat = 10 {
        didSet { setNeedsDisplay() }
    }

    @IBInspectable open var trackImage: UIImage? {
        didSet { setNeedsDisplay() }
    }

    @IBInspectable open var trackBackgroundColor: UIColor = UIColor.gray {
        didSet { setNeedsDisplay() }
    }

    @IBInspectable open var trackFillColor: UIColor = UIColor.blue {
        didSet { setNeedsDisplay() }
    }

    @IBInspectable open var trackBorderColor:UIColor = UIColor.clear {
        didSet { setNeedsDisplay() }
    }

    @IBInspectable open var trackBorderWidth: CGFloat = 0 {
        didSet { setNeedsDisplay() }
    }

    @IBInspectable open var centerFillColor: UIColor = UIColor.white {
        didSet { setNeedsDisplay() }
    }

    @IBInspectable open var centerImage: UIImage? {
        didSet { setNeedsDisplay() }
    }

    @IBInspectable open var contentView: UIView {
        return self.constants.contentView
    }

    required override public init(frame: CGRect) {
        super.init(frame: frame)
        internalInit()
        self.addSubview(contentView)
    }

    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        internalInit()
        self.addSubview(contentView)
    }

    func internalInit() {
        displayLink = CADisplayLink(target: self, selector: #selector(CircleProgressView.displayLinkTick))
        displayLink?.add(to: RunLoop.main, forMode: .defaultRunLoopMode)
        displayLink?.isPaused = true
    }

    override open func draw(_ rect: CGRect) {

        super.draw(rect)

        let innerRect = rect.insetBy(dx: trackBorderWidth, dy: trackBorderWidth)

        internalProgress = (internalProgress/1.0) == 0.0 ? constants.minimumValue : progress
        internalProgress = (internalProgress/1.0) == 1.0 ? constants.maximumValue : internalProgress
        internalProgress = clockwise ?
                                (-constants.twoSeventyDegrees + ((1.0 - internalProgress) * constants.circleDegress)) :
                                (constants.ninetyDegrees - ((1.0 - internalProgress) * constants.circleDegress))

        let context = UIGraphicsGetCurrentContext()

        // background Drawing
        let circlePath = UIBezierPath(ovalIn: CGRect(x: innerRect.minX, y: innerRect.minY, width: innerRect.width, height: innerRect.height))

        if trackBackgroundColor != UIColor.clear {
            trackBackgroundColor.setFill()
            circlePath.fill();
        }

        if trackBorderWidth > 0 {
            circlePath.lineWidth = trackBorderWidth
            trackBorderColor.setStroke()
            circlePath.stroke()
        }

        // progress Drawing
        let progressPath = UIBezierPath()
        let progressRect: CGRect = CGRect(x: innerRect.minX, y: innerRect.minY, width: innerRect.width, height: innerRect.height)
        let center = CGPoint(x: progressRect.midX, y: progressRect.midY)
        let radius = progressRect.width / 2.0
        let startAngle:CGFloat = clockwise ? CGFloat(-internalProgress * M_PI / 180.0) : CGFloat(constants.twoSeventyDegrees * M_PI / 180)
        let endAngle:CGFloat = clockwise ? CGFloat(constants.twoSeventyDegrees * M_PI / 180) : CGFloat(-internalProgress * M_PI / 180.0)

        progressPath.addArc(withCenter: center, radius:radius, startAngle:startAngle, endAngle:endAngle, clockwise:!clockwise)
        progressPath.addArc(withCenter: center, radius:radius-trackWidth, startAngle:endAngle, endAngle:startAngle, clockwise:clockwise)
        //progressPath.addLine(to: CGPoint(x: progressRect.midX, y: progressRect.midY))
        progressPath.close()

        context?.saveGState()

        progressPath.addClip()

        if let trackImage = trackImage {
            trackImage.draw(in: innerRect)
        } else if trackFillColor != UIColor.clear {
            trackFillColor.setFill()
            circlePath.fill()
        }

        context?.restoreGState()

        // center Drawing
        let centerPath = UIBezierPath(ovalIn: CGRect(x: innerRect.minX + trackWidth, y: innerRect.minY + trackWidth, width: innerRect.width - (2 * trackWidth), height: innerRect.height - (2 * trackWidth)))

        if centerFillColor != UIColor.clear {
            centerFillColor.setFill()
            centerPath.fill()
        }

        if let centerImage = centerImage {
            context?.saveGState()
            centerPath.addClip()
            centerImage.draw(in: rect)
            context?.restoreGState()
        } else {
            let layer = CAShapeLayer()
            layer.path = centerPath.cgPath
            contentView.layer.mask = layer
        }
    }

    //MARK: - Progress Update

    open func setProgress(_ newProgress: Double, animated: Bool) {

        if animated {
            destinationProgress = newProgress
            displayLink?.isPaused = false
        } else {
            progress = newProgress
            displayLink?.isPaused = true
        }
    }

    //MARK: - CADisplayLink Tick

    internal func displayLinkTick() {

        let renderTime = refreshRate.isZero ? displayLink!.duration : refreshRate

        if destinationProgress > progress {
            progress += renderTime
            if progress >= destinationProgress {
                progress = destinationProgress
            }
        }
        else if destinationProgress < progress {
            progress -= renderTime
            if progress <= destinationProgress {
                progress = destinationProgress
            }
        } else {
            displayLink?.isPaused = true
        }
    }


}
